appender.go

```diff
@@ -9,24 +9,46 @@ import (
 	"database/sql/driver"
 	"errors"
 	"fmt"
+	"reflect"
 	"time"
 	"unsafe"
 )
 
-// Appender holds the DuckDB appender. It allows to load bulk data into a DuckDB database.
+type LogicalTypeEnum int
+
+const (
+	PRIMITIVE LogicalTypeEnum = iota
+	LIST
+	STRUCT
+)
+
+type SetColumnValue func(a *Appender, columnInfo *ColumnInfo, rowIdx C.idx_t, val interface{})
+
+type ColumnInfo struct {
+	vector   C.duckdb_vector
+	function SetColumnValue
+
+	colType LogicalTypeEnum
+
+	// Only for nested types so that we can recursively store the child column info
+	fields      int // the number of fields in the struct
+	columnInfos []ColumnInfo
+}
+
+// Appender holds the DuckDB appender. It allows data bulk loading into a DuckDB database.
 type Appender struct {
 	c        *conn
 	schema   string
 	table    string
 	appender *C.duckdb_appender
 	closed   bool
 
-	currentRow       C.idx_t
 	chunks           []C.duckdb_data_chunk
-	chunkVectors     []C.duckdb_vector
-	currentChunkIdx  int
 	currentChunkSize C.idx_t
-	chunkTypes       *C.duckdb_logical_type
+	columnTypes      *C.duckdb_logical_type
+	columnTypesPtr   *[1073741823]C.duckdb_logical_type
```

Comment:
I had another look at this, specifically how we allocate these C-style arrays. I think your/our approach in `initializeColumnTypes` is very similar to your approach for creating the `STRUCT` child types. My trouble with the `STRUCT` child-type approach is allocating the memory through Go, which potentially can get garbage collected. I.e., you allocate the slice in Go but then pass its data into C, pretending it to be a C-array. Is it better to allocate the memory directly in C (`C.malloc`), as we do for the column types? I found [this](https://stackoverflow.com/questions/51525876/use-go-slice-in-c). 

Additionally, I prefer something less verbose than `*[1073741823]C.duckdb_logical_type`. I had a go at this; here is my (hacky) `diff.` It also destroys the logical types for the `STRUCT` child type creation and omits the dummy logical type.

```diff
@@ -45,8 +48,8 @@ type Appender struct {
 
-       columnTypes      *C.duckdb_logical_type
-       columnTypesPtr   *[1073741823]C.duckdb_logical_type
+       columnTypes      []C.duckdb_logical_type
+       columnTypesPtr   unsafe.Pointer

@@ -265,27 +271,26 @@ func (a *Appender) InitializeColumnTypesAndInfos(v reflect.Type, colIdx int) (C.
                }
 
                structType := v
-
                columnInfo := ColumnInfo{function: f, colType: STRUCT, columnInfos: make([]ColumnInfo, v.NumField()), fields: v.NumField()}
 
-               // Create an array of the struct fields TYPES
-               // and an array of the struct fields NAMES
-               fieldTypes := make([]C.duckdb_logical_type, structType.NumField())
-               fieldNames := make([]*C.char, structType.NumField())
+               // Create an array of the struct field TYPES
+               // and an array of the struct field NAMES
+               fieldTypesPtr, fieldTypes := mallocLogicalTypeSlice(structType.NumField())
+               fieldNamesPtr, fieldNames := mallocCStringSlice(structType.NumField())
                for i := 0; i < structType.NumField(); i++ {
                        fieldTypes[i], columnInfo.columnInfos[i] = a.InitializeColumnTypesAndInfos(structType.Field(i).Type, i)
                        fieldNames[i] = C.CString(structType.Field(i).Name)
                }
 
-               fieldTypesPtr := (*C.duckdb_logical_type)(unsafe.Pointer(&fieldTypes[0]))
-               fieldNamesPtr := (**C.char)(unsafe.Pointer(&fieldNames[0]))
-
-               structLogicalType := C.duckdb_create_struct_type(fieldTypesPtr, fieldNamesPtr, C.idx_t(structType.NumField()))
+               structLogicalType := C.duckdb_create_struct_type((*C.duckdb_logical_type)(fieldTypesPtr),
+                       (**C.char)(fieldNamesPtr), C.idx_t(structType.NumField()))
 
                for i := 0; i < structType.NumField(); i++ {
+                       C.duckdb_destroy_logical_type(&fieldTypes[i])
                        C.free(unsafe.Pointer(fieldNames[i]))
-                       C.free(unsafe.Pointer(fieldTypes[i]))
                }
+               C.free(fieldTypesPtr)
+               C.free(fieldNamesPtr)

@@ -297,23 +302,18 @@
-// Create an array of duckdb types from a list of go types
+// initializeColumnTypes creates an array of DuckDB types from a list of go types
 func (a *Appender) initializeColumnTypes(args []driver.Value) {
        a.columnInfos = make([]ColumnInfo, len(args))
-       defaultLogicalType := C.duckdb_create_logical_type(0)
-       columnTypes := C.malloc(C.size_t(len(args)) * C.size_t(unsafe.Sizeof(defaultLogicalType)))
-
-       a.columnTypesPtr = (*[1<<30 - 1]C.duckdb_logical_type)(columnTypes)
+       a.columnTypesPtr, a.columnTypes = mallocLogicalTypeSlice(len(args))
 
        for i, val := range args {
                if val == nil {
                        panic(fmt.Sprintf("The first row cannot contain null values (column %d)", i))
                }
                v := reflect.ValueOf(val)
-               a.columnTypesPtr[i], a.columnInfos[i] = a.InitializeColumnTypesAndInfos(v.Type(), i)
+               a.columnTypes[i], a.columnInfos[i] = a.InitializeColumnTypesAndInfos(v.Type(), i)
        }
-
-       a.columnTypes = (*C.duckdb_logical_type)(columnTypes)
 }
 
 func (c *ColumnInfo) getChildVectors(vector C.duckdb_vector) error {

@@ -342,8 +342,8 @@  
 func (a *Appender) addChunk(colCount int) error {
        a.currentChunkSize = 0
-       // duckdb_create_data_chunk takes an array of duckdb_logical_type and a column count
-       dataChunk := C.duckdb_create_data_chunk(a.columnTypes, C.idx_t(colCount))
+       columnTypesPtr := (*C.duckdb_logical_type)(a.columnTypesPtr)
+       dataChunk := C.duckdb_create_data_chunk(columnTypesPtr, C.idx_t(colCount))
        C.duckdb_data_chunk_set_size(dataChunk, C.duckdb_vector_size())
 
        for i := 0; i < colCount; i++ {

@@ -469,11 +469,29 @@ func (a *Appender) appendChunks() error {
 
        // free the column types
        for i := range a.columnInfos {
-               C.duckdb_destroy_logical_type(&a.columnTypesPtr[i])
+               C.duckdb_destroy_logical_type(&a.columnTypes[i])
        }
 
-       C.free(unsafe.Pointer(a.columnTypes))
+       C.free(a.columnTypesPtr)
        return nil
 }
```

And here are the two new functions that I added.
```go
func mallocCStringSlice(count int) (unsafe.Pointer, []*C.char) {
	var dummyExpr *C.char
	size := C.size_t(unsafe.Sizeof(dummyExpr))

	cStringsPtr := unsafe.Pointer(C.malloc(C.size_t(count) * size))
	cStrings := (*[1 << 30]*C.char)(cStringsPtr)[:count:count]
	return cStringsPtr, cStrings
}

func mallocLogicalTypeSlice(count int) (unsafe.Pointer, []C.duckdb_logical_type) {
	var dummyType C.duckdb_logical_type
	size := C.size_t(unsafe.Sizeof(dummyType))

	columnTypesPtr := unsafe.Pointer(C.malloc(C.size_t(count) * size))
	columnTypes := (*[1 << 30]C.duckdb_logical_type)(columnTypesPtr)[:count:count]
	return columnTypesPtr, columnTypes
}
```